src/azure-cli/azure/cli/command_modules/resource/_packing_engine.py

```diff
@@ -0,0 +1,188 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import os
+import json
+from knack.util import CLIError
+from azure.cli.core.util import read_file_content
+from azure.cli.command_modules.resource.custom import _remove_comments_from_json
+from azure.cli.core.profiles import ResourceType, get_sdk
+
+
+class PackagedTemplate():  # pylint: disable=too-few-public-methods
+    def __init__(self, template, artifacts):
+        self.RootTemplate = template
+        self.Artifacts = artifacts
+
+
+class PackingContext():  # pylint: disable=too-few-public-methods
+    def __init__(self, root_template_directory):
+        self.RootTemplateDirectory = os.path.abspath(root_template_directory)
+        self.CurrentDirectory = os.path.abspath(root_template_directory)
+        self.Artifact = []
+
+
+# Packs the specified template and its referenced artifacts for use in a Template Spec.
+# :param template_file: The path to the template spec .json file.
+# :type name: str
+
+def pack(cmd, template_file):
+    root_template_file_path = os.path.abspath(template_file)
+    context = PackingContext(os.path.dirname(root_template_file_path))
+    template_content = read_file_content(template_file)
+    sanitized_template = _remove_comments_from_json(template_content)
+    template_json = json.loads(json.dumps(sanitized_template))
+    _pack_artifacts(cmd, root_template_file_path, context)
+    return PackagedTemplate(template_json, getattr(context, 'Artifact'))
+
+#  Recursively packs the specified template and its referenced artifacts and
+#  adds the artifacts to the current packing context.
+
+# :param template_abs_file_path: The path to the template spec .json file to pack.
+# :type template_abs_file_path : str
+# :param context : The packing context of the current packing operation
+# :type content : PackingContext
+# :param artifactableTemplateObj : The packagable template object
+# :type artifactableTemplateObj : JSON
+
+
+def _pack_artifacts(cmd, template_abs_file_path, context):
+    originalDirectory = getattr(context, 'CurrentDirectory')
+    try:
+        context.CurrentDirectory = os.path.dirname(template_abs_file_path)
+        template_content = read_file_content(template_abs_file_path)
+        artifactableTemplateObj = sanitized_template = _remove_comments_from_json(template_content)
+        template_json = json.loads(json.dumps(sanitized_template))
+        templateLinktoArtifactObjs = _get_template_links_to_artifacts(cmd, artifactableTemplateObj, includeNested=True)
```

Comment:
nit: Prefer `snake_case` over `camelCase`. See [PEP 8: Function and Variable Names](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names) (this does not block the PR).